FINAL

turchenkoalex opened the pull request.

turchenkoalex committed:
Initial commit (test 1)

turchenkoalex committed:
Внесены мелкие изменения

Вынесен код генерации случайного символа в отдельную функцию.
Заменен вызов формирования строки для вывода на печать класса Cell.
Переименованы DNATypes на более логичные DNASymbols.
Переименованы файлы с категорией на Cell+Mutator

turchenkoalex commented on osx-1/main.m:
```diff
@@ -0,0 +1,34 @@
+//
+
+#import <Foundation/Foundation.h>
+#import "Core/Cell.h"
```
Подскажите: Если у нас есть категория какого то класс и мы ей пользуемся, и импортируем заголовочные файлы категории, то нужно ли (принято ли это) при этом делать импорт заголовочных файлов основного класса над которым создана категория?

unfilet replied:
Нет, так как он уже подключен в файле категории. Файл подключается только один раз.

slebedev replied:
Но замечу, что ошибки не будет, в отличии от "сишного" `#include`. Дело в том, что `#import` сначала проверяет, не импортировался ли этот файл ранее, и не будет лишний раз его импортировать.
Материал первой лекции, вроде бы.

turchenkoalex replied:
Да, я понимаю как это работает. Собственно вопрос был не в том можно ли так или иначе, а в том как это принято в Objective C.

ekrymov commented:
Запускаю проект у себя. Каждый раз генерируются одинаковые массивы, соответственно и результат всегда одинаковый: 80, 82 (с параметрами из проекта).

turchenkoalex commented:
Для генерации псевдослучайных чисел использовалась функция random(), вестимо она себя ведет таким образом (не могу в данный момент точно сказать так как не силен в "c"), что выдает псевдослучайные числа в одной и той же последовательности, но тем паче, проверять намного проще (так как в условиях задачи на случайность никаких ограничений не накладывалось)

ekrymov commented:
То, что в условиях написано "случайным образом" уже говорит о том, что клетки доолжны быть постоянно разные. Для решения проблемы замени генератор рандомных чисел на arc4random()

slebedev commented:
Во-первых, для использования функции random её нужно сначала, как бы, проинициализировать, задать отправную точку. Делается это с помощью функции srandom.

То, что значения одинаковые после генератора можно объяснить, так как генератор random - генератор псевдослучайных чисел, и если он каждый раз будет помещён в одни и те же условия, то наши "случайные" значения будут повторяться.

Есть, скажем так, более "случайная" функция arc4random(); Ей не нужно задавать начальное состояние. Так же есть arc4random_uniform()

turchenkoalex commented:
Да спасибо. Я так понимаю что будет достаточно выполнить srandom при старте приложения, для исправления условной проблемы.
Так как это все таки тестовое задание, а не production код, то мне показалось, достаточно писать так что бы было понятно что "ты" понимаешь принципы построения кода, а не перфекционистский код. В моем коде есть еще проблемы с опять таки случайными числами, например вероятности мутации гена не распределены равномерно, а один из генов будет иметь 50% вероятность выпадения нежели остальные. Да можно потратить еще час и сделать равно по 33%, но особого смысла в этом не вижу.

slebedev commented:
Могу вам так сказать, современные процессоры настолько быстры, что значение времени в обоих случаях будет одно и то же, и начальное состояние будет опять же, одинаковым. 

Понятное дело, что вы учитесь. Но понимайте все комментарии правильно - это путь получения информации, для понимания того, что вы делаете. В дискуссиях вы сможете как улучшать свой код, так и свои знания и навыки.

turchenkoalex commented:
Совершенно с вами согласен. Для себя я усвоил то о чем вы мне сказали. Потому то мы и проходим обучение совместно, а не ковыряемся в коде в одиночку.

turchenkoalex committed:
FINAL: Изменения в методе мутации

Упрощен метод мутации: сделан на основе перемешивания массива индексов
Заменен random() на arc4random()
Убран import Core/Cell.h
